### PR TITLE
Add genesis block's components

### PR DESCRIPTION
**Pull request by MizukiSonoko**
### Usage Examples or Tests *[optional]*

`cargo run -- --genesis tests/genesis.json`

### CODE REVIEW

File: iroha/src/genesis.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    fn load_genesis_file_from_path<P: AsRef<Path> + Debug>(
+        path: P,
+        init_config: &InitConfiguration,
+    ) -> Result<GenesisBlock, String> {
+        let file =
+            File::open(path).map_err(|e| format!("Failed to open a genesis block file: {}", e))?;
+        let reader = BufReader::new(file);
+        let raw_block: RawGenesisBlock = serde_json::from_reader(reader)
+            .map_err(|e| format!("Failed to deserialize json from reader: {}", e))?;
+        let genesis_key_pair = KeyPair {
```
Comment (MizukiSonoko) [inline iroha/src/genesis.rs:53]: Should I make constructor of KeyPair?

File: iroha/config.json [hunk 2]
```diff
@@ -43,6 +43,9 @@
     "digest_function": "ed25519",
     "payload": "9ac47abf59b356e0bd7dcbbbb4dec080e302156a48ca907e47cb6aea1d32719e7233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
   },
+  "GENESIS_CONFIGURATION": {
+    "GENESIS_FILE_PATH": "tests/genesis.json"
+  },
```
Comment (eadventurous) [inline iroha/config.json:48]: We already pass genesis file path in the command line through `./iroha --genesis="tests/genesis.json"`, so why should we need this field here?

File: iroha/src/main.rs [hunk 3]
```diff
@@ -1,13 +1,35 @@
...
+                .required(false),
+        )
+        .get_matches();
+
     let mut configuration = Configuration::from_path(CONFIGURATION_PATH)?;
     configuration.load_environment()?;
+
+    let genesis_path_option = matches.value_of(GENESIS);
+    if let Some(genesis_path) = genesis_path_option {
+        println!("load genesis block");
```
Comment (eadventurous) [inline iroha/src/main.rs:29]: `println!("Loading genesis block from the path: {}", genesis_path);`

File: iroha/tests/test_config.json [hunk 4]
```diff
@@ -8,7 +8,7 @@
     "COMMIT_TIME_MS": 1000,
     "TX_RECEIPT_TIME_MS": 100,
     "MAX_FAULTY_PEERS": 1,
-    "TRUSTED_PEERS": [
+    "IROHA_TRUSTED_PEERS": [
```
Comment (eadventurous) [inline iroha/tests/test_config.json:11]: `IROHA_TRUSTED_PEERS` is the name of the environment variable, but the name of the field is just `TRUSTED_PEERS`. So as far as I remember the config is correct here.

File: iroha/src/lib.rs [hunk 5]
```diff
@@ -99,6 +101,7 @@ pub struct Iroha {
     block_sync_message_receiver: Arc<RwLock<BlockSyncMessageReceiver>>,
     world_state_view: Arc<RwLock<WorldStateView>>,
     block_sync: Arc<RwLock<BlockSynchronizer>>,
+    genesis_block: Arc<Option<GenesisBlock>>,
```
Comment (eadventurous) [inline iroha/src/lib.rs:104]: I think Genesis Block does not need to be wrapped with Arc (Atomic Reference Counter), it is not passed into any threads or async functions, as far as I see.

File: iroha/src/genesis.rs (hunk 1 above)
Comment (eadventurous) [reply iroha/src/genesis.rs:53]: No I don't think we need constructor for the `KeyPair`, it would simply fill the fields...

File: iroha/src/genesis.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    impl GenesisConfiguration {
+        /// Load environment variables and replace predefined parameters with these variables
+        /// values.
+        pub fn load_environment(&mut self) -> Result<(), String> {
+            if let Ok(genesis_file_path) = env::var(GENESIS_FILE_PATH) {
+                self.genesis_file_path = genesis_file_path;
+            }
+            Ok(())
+        }
```
Comment (eadventurous) [inline iroha/src/genesis.rs:104]: We don't need to load genesis path from the environment, we simply tell in CLI if we want genesis to be loaded from path or not.

File: iroha/src/genesis.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        };
+        Ok(GenesisBlock {
+            txs: raw_block
+                .txs
+                .iter()
+                .map(|rtx| {
+                    Transaction::new(
+                        rtx.isi.clone(),
+                        <Account as Identifiable>::Id::genesis_account(),
+                        0,
```
Comment (eadventurous) [inline iroha/src/genesis.rs:68]: Please specify the default time to live here, it can be found in `tx.rs`. If it is zero, then the tx will be canceled immediately.

File: iroha/tests/genesis.json [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            "DomainName": "global"
+          }
+        }
+      },
+      {
+        "Register": {
+          "object": {
+            "Account": {
+              "id": {
+                "name": "root",
```
Comment (eadventurous) [inline iroha/tests/genesis.json:40]: Please change a name to `alice` (or actually anything else than `root`). Root was a bit of a special case we had, when we had a previous permission system, it is not needed anymore.

File: iroha/tests/genesis.json [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+              "trusted_peers_ids": [],
+              "triggers": [],
+              "parameters": []
+            }
+          },
+          "destination_id": {
+            "DomainName": "global"
+          }
+        }
+      },
```
Comment (eadventurous) [inline iroha/tests/genesis.json:24]: Please remove Register Peer, it might break consensus if in genesis, it is an edge case I will need to check later.

File: iroha/src/genesis.rs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        )
+    }
+
+    /// This method will build `Configuration` from a json *pretty* formatted file (without `:` in
+    /// key names).
+    /// # Panics
+    /// This method will panic if configuration file presented, but has incorrect scheme or format.
+    /// # Errors
+    /// This method will return error if system will fail to find a file or read it's content.
+    fn load_genesis_file_from_path<P: AsRef<Path> + Debug>(
```
Comment (eadventurous) [inline iroha/src/genesis.rs:44]: It is better to combine this two functions into one, it is actually just one function logically.

File: iroha/src/genesis.rs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+#[derive(Clone, Deserialize, Debug)]
+struct RawGenesisBlock {
+    pub txs: Vec<GenesisTx>,
+}
+
+/// GenesisTx is a transaction for inisialize settings.
+#[derive(Clone, Deserialize, Debug)]
+struct GenesisTx {
```
Comment (eadventurous) [inline iroha/src/genesis.rs:22]: An idiomatic way to implement a conversion in rust would be to implement: `impl TryFrom<GenesisTx> for AcceptedTransaction` and use it in the `map` below.

**Review by eadventurous — CHANGES_REQUESTED** [review]

**Comment by eadventurous** [comment]
I would also suggest renaming all `tx` simply to `transaction`. It is just a convention we use here, to keep the naming cohesive.

File: iroha/src/genesis.rs (hunk 10 above)
Comment (MizukiSonoko) [reply iroha/src/genesis.rs:44]: ok!

File: iroha/src/genesis.rs (hunk 1 above)
Comment (MizukiSonoko) [reply iroha/src/genesis.rs:53]: ok!

File: iroha/src/genesis.rs (hunk 7 above)
Comment (MizukiSonoko) [reply iroha/src/genesis.rs:68]: ok!

File: iroha/tests/genesis.json (hunk 9 above)
Comment (MizukiSonoko) [reply iroha/tests/genesis.json:24]: ok!

File: iroha/tests/genesis.json (hunk 8 above)
Comment (MizukiSonoko) [reply iroha/tests/genesis.json:40]: ok!

File: iroha/config.json (hunk 2 above)
Comment (MizukiSonoko) [reply iroha/config.json:48]: Sorry, I forgot to remove this.

File: iroha/config.json [hunk 12]
```diff
@@ -11,20 +11,20 @@
...
-        "address": "localhost:1340",
-        "public_key": "ed20faca9e8aa83225cb4d16d67f27dd4f93fc30ffa11adc1f5c88fd5495ecc91020"
+        "address": "195.162.0.1:23",
+        "public_key": "ed207233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
       },
       {
-        "address": "localhost:1340",
-        "public_key": "ed208e351a70b6a603ed285d666b8d689b680865913ba03ce29fb7d13a166c4e7f1f"
+        "address": "195.162.0.1:231",
+        "public_key": "ed207233bfc89dcbd68c19fde6ce6158225298ec1131b6a130d1aeb454c1ab5183c0"
```
[the remark this answers is not in the document]
Comment (MizukiSonoko) [reply iroha/config.json:27]: Sorry, maybe I merged an invalid commit by mistake.

File: iroha/src/genesis.rs (hunk 6 above)
Comment (MizukiSonoko) [reply iroha/src/genesis.rs:104]: ok!

File: iroha/src/genesis.rs (hunk 11 above)
Comment (MizukiSonoko) [reply iroha/src/genesis.rs:22]: I agree with your idea, but I have a question.
Now, conversion logic requires keypair for signing....
so how to use TryFrom? I thought some method but not I'm not convinced.
1) GenesisTx has a keypair
2) Define PairType as follow
```
struct GenesisTxWithKey {
    txs: GenesisTx
    key: KeyPair
}

impl TryFrom<GenesisTxWithKey> for AcceptedTransaction
```

**Review by MizukiSonoko — COMMENTED** [review]
Thanks for ReView
(sorry for this PR did contain some wrongs)

File: iroha/src/genesis.rs (hunk 11 above)
Comment (eadventurous) [reply iroha/src/genesis.rs:22]: Okay, I see. Yes you are right, implementing `TryFrom` will not be good here. Maybe you can do something like that:
```rust
impl GenesisTransaction {
    pub fn sign_and_accept(key_pair) -> AcceptedTransaction {
        ...
    }
}
```

File: iroha/src/config.rs [hunk 13]
```diff
@@ -35,6 +35,8 @@ pub struct Configuration {
     pub logger_configuration: LoggerConfiguration,
     /// Configuration for initial setup
     pub init_configuration: InitConfiguration,
+    /// `GenesisBlock` path.
+    pub genesis_block_path: Option<String>,
```
Comment (MizukiSonoko) [inline iroha/src/config.rs:39]: Should I use `Option<PathBuf>`? 
In my option, unnecessary mutable path...

Comment (eadventurous) [reply iroha/src/config.rs:39]: Yes, sure, you can use path buff

**Review by eadventurous — APPROVED** [review]
